colossalai/inference/tensor_parallel/modeling/llama.py

```diff
@@ -237,16 +249,16 @@ def llama_flash_attn_kvcache_forward(
         key_states = self.k_proj(hidden_states).view(bsz, q_len, self.num_heads, self.head_dim)
         key_states_transposed = key_states.transpose(1, 2)
         value_states = self.v_proj(hidden_states).view(bsz, q_len, self.num_heads, self.head_dim)
-        
+
         # cos, sin = self.rotary_emb(value_states_transposed, seq_len=kv_seq_len)
-        cos ,sin = infer_state.position_cos, infer_state.position_sin
-    
-        cos_sin_cache = torch.cat((cos, sin), dim=-1)
-        
-        from vllm.pos_encoding_ops import rotary_embedding_neox
-        
-        rotary_embedding_neox(position_ids, query_states, key_states_transposed, self.head_dim, cos_sin_cache)
-        
+        cos, sin = infer_state.position_cos, infer_state.position_sin
+
+        if HAS_VLLM_KERNERL:
+            cos_sin_cache = torch.cat((cos, sin), dim=-1)
+            rotary_embedding_neox(position_ids, query_states, key_states_transposed, self.head_dim, cos_sin_cache)
+        else:
+            query_states, key_states = apply_rotary_pos_emb(query_states, key_states_transposed, cos, sin, position_ids)
```

Comment:
Can apply_rotary_pos_emb  be found here？

Reply:
import.